---
title: "R Assignment"
author: Jeniffer Perea-Lopez
output: html_notebook
date: 03-13-2023
---

```{r}
#Load applicable packages
library(tidyverse)
library(dplyr)
library(ggplot2)
library(gtools)
```

```{r}
#load text files "fang_et_al_genotypes.txt" and "snp_position.txt"; pull from GitHub
fang_et_al <- read_tsv("https://github.com/EEOB-BioData/BCB546_Spring2023/raw/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt")
snp_position <- read_tsv("https://github.com/EEOB-BioData/BCB546_Spring2023/raw/main/assignments/UNIX_Assignment/snp_position.txt")
```

Part I

Data Inspection

1. fang_et_al_genotypes.txt

```{r}
#Analyze structure and dimensions of file:

dim(fang_et_al)
#Dimensions: 2782x986
ncol(fang_et_al)
#number of columns: 986
nrow(fang_et_al)
#number of rows: 2782
head(fang_et_al)
#Obtain first 6 observations of data set
tail(fang_et_al)
#Obtain last 6 observations of data set
names(fang_et_al)
#Return column headers
```

2. snp_position.txt

```{r}
#Analyze structure and dimensions of file:

dim(snp_position)
#Dimensions: 983x15
ncol(snp_position)
#number of columns: 15
nrow(snp_position)
#number of rows: 983
head(snp_position)
#Obtain first 6 observations of data set
tail(snp_position)
#Obtain last 6 observations of data set
names(snp_position)
#Return column headers
```

Data Processing

```{r}
#Subsetting maize data (Group = ZMMIL, ZMMLR and ZMMMR)
maize <- filter(fang_et_al, Group %in% c("ZMMIL","ZMMLR","ZMMMR"))
#Remove extraneous information from maize, which includes columns 1-3
maize <- maize[-c(2:3)]
#Transpose maize file
maize_transposed <- t(maize)
#Set "Sample_ID" row as column names
colnames(maize_transposed) <- maize_transposed[1,] 
maize_transposed <- maize_transposed[-1,]

#Subsetting teosinte data (Group = ZMPBA, ZMPIL and ZMPJA)
teosinte <- filter(fang_et_al, Group %in% c("ZMPBA","ZMPIL","ZMPJA"))
#Remove extraneous information from teosinte, which includes columns 1-3
teosinte <- teosinte[-c(2:3)]
#Transpose teosinte file 
teosinte_transposed <- t(teosinte)
#Set "Sample_ID" row as column names
colnames(teosinte_transposed) <- teosinte_transposed[1,] 
teosinte_transposed <- teosinte_transposed[-1,]

#Remove extraenous information from snp_position, which includes columns 2 and 5-15, keeping only "SNP_ID," "Chromosome" and "Position"
snp_position <- snp_position[c(1,3,4)]

#Merge transposed files with snp_position 
maize_snp <-  merge(snp_position,maize_transposed,by.x="SNP_ID", by.y="row.names")
teosinte_snp <-  merge(snp_position,teosinte_transposed,by.x = "SNP_ID",by.y = "row.names")

#FOR MAIZE: Sort data for each chromosome with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?

dir.create("maize")

for (i in 1:10){
  #filter Chromosome "i" data in maize_snp based on the "Chromosome" column
  maize_chr <- filter(maize_snp,Chromosome %in% c(i))
  #Sort generated files based on the "Position" column in increasing order; convert "Position" column to character, then to numeric
  maize_chr_increasing <- maize_chr[order(as.numeric(as.character(maize_chr$Position))),]
  #Create corresponding files based on increasing position and Chromosome "i"
  maize_increasing <- paste("maize/maize_increasing_chr",i,".txt", sep="")
  #Export data frame to files corresponding to each Chromosome "i" and based on increasing position values
  write.table(maize_chr_increasing,file= maize_increasing, row.names = FALSE, sep = "\t", col.names = TRUE, quote = FALSE)
  
#Sort data for each chromosome with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -
  
  #Sort maize_chr files based on the "Position" column in decreasing order
  maize_chr_decreasing_position <- mixedsort(as.character(maize_chr$Position), decreasing = TRUE)
  maize_chr_decreasing <- maize_chr
  maize_chr_decreasing$Position <- maize_chr_decreasing_position
  #Replace "?" with "-"
  maize_chr_decreasing[maize_chr_decreasing=="?/?"] <- "-/-"
  #Create corresponding files based on decreasing position and Chromosome "i"
  maize_decreasing <- paste("maize/maize_decreasing_chr",i,".txt", sep="")
  #Export data frame to files corresponding to each Chromosome "i" and based on decreasing position values
  write.table(maize_chr_decreasing,file= maize_decreasing, row.names = FALSE, sep = "\t", col.names = TRUE, quote = FALSE)
}

#FOR TEOSINTE: Sort data for each chromosome with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?

dir.create("teosinte")

for (i in 1:10){
  #filter Chromosome "i" data in teosinte_snp based on the "Chromosome" column
  teosinte_chr <- filter(teosinte_snp,Chromosome %in% c(i))
  #Sort generated files based on the "Position" column in increasing order; convert "Position" column to character, then to numeric
  teosinte_chr_increasing <- teosinte_chr[order(as.numeric(as.character(teosinte_chr$Position))),]
  #Create corresponding files based on increasing position and Chromosome "i"
  teosinte_increasing <- paste("teosinte/teosinte_increasing_chr",i,".txt", sep="")
  #Export data frame to files corresponding to each Chromosome "i" and based on increasing position values
  write.table(teosinte_chr_increasing,file= teosinte_increasing, row.names = FALSE, sep = "\t", col.names = TRUE, quote = FALSE)
  
#Sort data for each chromosome with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -
  
  #Sort teosinte_chr files based on the "Position" column in decreasing order
    teosinte_chr_decreasing_position <- mixedsort(as.character(teosinte_chr$Position), decreasing = TRUE)
  teosinte_chr_decreasing <- teosinte_chr
  teosinte_chr_decreasing$Position <- teosinte_chr_decreasing_position
  #Replace "?" with "-"
  teosinte_chr_decreasing[teosinte_chr_decreasing=="?/?"] <- "-/-"
  #Create corresponding files based on decreasing position and Chromosome "i"
  teosinte_decreasing <- paste("teosinte/teosinte_decreasing_chr",i,".txt", sep="")
  #Export data frame to files corresponding to each Chromosome "i" and based on decreasing position values
  write.table(teosinte_chr_decreasing,file= teosinte_decreasing, row.names = FALSE, sep = "\t", col.names = TRUE, quote = FALSE)
}

```

Part II: Visualization

```{r}
#SNPs per chromosome

#Subsetting maize (Group = ZMMIL, ZMMLR and ZMMMR) and teosinte (Group = ZMPBA, ZMPIL and ZMPJA) data
maize_teosinte <- filter(fang_et_al,Group %in% c("ZMMIL","ZMMLR","ZMMMR","ZMPBA","ZMPIL","ZMPJA"))
maize_teosinte <- maize_teosinte[-c(2)]
maize_teosinte_transposed <- t(maize_teosinte)
colnames(maize_teosinte_transposed) <- maize_teosinte_transposed[1,] 
maize_teosinte_transposed <- maize_teosinte_transposed[-1,]
maize_teosinte_snp <- merge(snp_position,maize_teosinte_transposed,by.x = "SNP_ID", by.y = "row.names")

#Plot SNPs per chromosome 
ggplot (data = maize_teosinte_snp) + geom_bar(mapping = aes(x = Chromosome, fill = Chromosome)) +xlab(label = "Chromosome") + ylab(label = "SNPs") + ggtitle("SNPs per Chromosome") + scale_x_discrete(limit=c(1:10, "unknown", "multiple"))
#Chromosome 1 has the most SNPs, whereas chromosome 10 has the least amount of SNPs

#Plot SNPs per chromosome for both maize and teosinte individually 

#FOR MAIZE:
ggplot (data = maize_snp) + geom_bar(mapping = aes(x = Chromosome, fill = Chromosome)) +xlab(label = "Chromosome") + ylab(label = "SNPs") + ggtitle("SNPs per Chromosome for Maize") + scale_x_discrete(limit=c(1:10, "unknown", "multiple"))

#FOR TEOSINTE:
ggplot (data = teosinte_snp) + geom_bar(mapping = aes(x = Chromosome, fill = Chromosome)) +xlab(label = "Chromosome") + ylab(label = "SNPs") + ggtitle("SNPs per Chromosome for Teosinte") + scale_x_discrete(limit=c(1:10, "unknown", "multiple"))
#From the plots, there does not seem to be a difference between maize and teosinte in terms of amount of SNP positions found in these two genotypes

#Plot distribution of SNPs across each chromosome for both maize and teosinte individually

#FOR MAIZE: 
#Count the number of SNPs for each SNP_ID 
count_snps_maize <- rowSums(maize_snp == "A/G" | maize_snp == "A/C" | maize_snp == "A/T" |  maize_snp == "C/G" | maize_snp == "G/T" | maize_snp == "C/T" | maize_snp == "A/A" | maize_snp == "C/C" | maize_snp == "G/G" | maize_snp == "T/T") 
#Create a dataframe with four columns: "SNP_ID_maize," "Chromosome_maize," "Position_maize," "Density_maize"
SNP_ID_maize <- maize_snp$SNP_ID
Chromosome_maize <- maize_snp$Chromosome
Position_maize <- as.numeric(maize_snp$Position)
Density_maize <- as.numeric(count_snps_maize)
maize_snp_count_data <- data.frame(SNP_ID_maize, Chromosome_maize, Position_maize, Density_maize)
#Plot SNPs across chromosome for maize
ggplot(maize_snp_count_data, aes(x=Position_maize, y=Density_maize)) + geom_point(size = 1, color = Chromosome_maize) + facet_wrap(~Chromosome_maize) + labs(title = "Distribution of SNPs across each Chromosome for Maize", x = "Position", y = "Count SNPs") + scale_x_continuous(limits = c(100000,300000000))

#FOR TEOSINTE: 
#Count the number of SNPs for each SNP_ID 
count_snps_teosinte <- rowSums(teosinte_snp == "A/G" | teosinte_snp == "A/C" | teosinte_snp == "A/T" |  teosinte_snp == "C/G" | teosinte_snp == "G/T" | teosinte_snp == "C/T" | teosinte_snp == "A/A" | teosinte_snp == "C/C" | teosinte_snp == "G/G" | teosinte_snp == "T/T") 
#Create a dataframe with four columns: "SNP_ID_teosinte," "Chromosome_teosinte," "Position_teosinte," "Density_teosinte"
SNP_ID_teosinte <- teosinte_snp$SNP_ID
Chromosome_teosinte <- teosinte_snp$Chromosome
Position_teosinte <- as.numeric(teosinte_snp$Position)
Density_teosinte <- as.numeric(count_snps_teosinte)
teosinte_snp_count_data <- data.frame(SNP_ID_teosinte, Chromosome_teosinte, Position_teosinte, Density_teosinte)
#Plot SNPs across chromosome for teosinte
ggplot(teosinte_snp_count_data, aes(x=Position_teosinte, y=Density_teosinte)) + geom_point(size = 1, color = Chromosome_teosinte) + facet_wrap(~Chromosome_teosinte) + labs(title = "Distribution of SNPs across each Chromosome for Teosinte", x = "Position", y = "Count SNPs") + scale_x_continuous(limits = c(100000,300000000))

#Missing data and amount of heterozygosity 

#Create a data frame with seven columns: "Group" and "Sample_ID" for both maize and teosinte, "Category" to distinguish between maize and teosinte, "Homozygous" for number of homozygous sites, "Heterozygous" for number of heterozygous sites, and "Missing_Data" for amount of missing data of each sample
Group <- maize_teosinte$Group
Sample_ID <- maize_teosinte$Sample_ID
Homozygous <- rowSums(maize_teosinte == "A/A" | maize_teosinte == "C/C" | maize_teosinte == "G/G" | maize_teosinte == "T/T")
Heterozygous <- rowSums(maize_teosinte == "A/G" | maize_teosinte == "A/C" | maize_teosinte == "A/T" | maize_teosinte == "C/G" | maize_teosinte == "G/T" | maize_teosinte == "C/T")
Missing_Data <- rowSums(maize_teosinte == "?/?")
rep_group = c('ZMMIL' = 'Maize', 'ZMMLR' = 'Maize', 'ZMMMR' = 'Maize', 'ZMPBA' = 'Teosinte', 'ZMPIL' = "Teosinte", 'ZMPJA' = 'Teosinte' )
fang_mt <- filter(fang_et_al,Group %in% c("ZMMIL","ZMMLR","ZMMMR","ZMPBA","ZMPIL","ZMPJA"))
fang_mt$Group <- str_replace_all(fang_mt$Group, rep_group)
colnames(fang_mt)[3] = "Category"
Category <- fang_mt$Category
Homo_Hetero <- Homozygous/Heterozygous
Missing_Percent <- (Missing_Data/(Heterozygous + Homozygous + Missing_Data))*100
Homo_Percent <- (Homozygous/(Heterozygous + Homozygous + Missing_Data))*100
Hetero_Percent <- (Heterozygous/(Heterozygous + Homozygous + Missing_Data))*100
dataframe_maize_teosinte <- data.frame(Sample_ID, Group, Category, Homozygous, Homo_Percent, Heterozygous, Hetero_Percent, Homo_Hetero, Missing_Data, Missing_Percent)

#Plot number of homozygous sites of each sample by group 
ggplot(dataframe_maize_teosinte) + geom_point(mapping = aes(x=Sample_ID, y=Homozygous, color=Group)) + labs(title = "Amount of Homozygous Sites by Group", x = "Sample_ID", y = "Amount of Homozygosity")
#Plot number of homozygous site of each sample by category
ggplot(dataframe_maize_teosinte) + geom_point(mapping = aes(x=Sample_ID, y=Homozygous, color=Category)) + labs(title = "Amount of Homozygous Sites by Category", x = "Sample_ID", y = "Amount of Homozygosity")

#Plot percentage of homozgyous sites of each sample by group
ggplot(data = dataframe_maize_teosinte, aes(x = Sample_ID, y = Homo_Percent, fill = Group )) + geom_bar(stat = "identity",  width = 0.4, position=position_dodge(width = 0.5)) + theme_classic() + labs(title = "Percentage of Homozygous Sites by Group", x = "Sample_ID", y = "Percent Homozygosity")
#Plot percentage of homozgyous sites of each sample by category
ggplot(data = dataframe_maize_teosinte, aes(x = Sample_ID, y = Homo_Percent, fill = Category )) + geom_bar(stat = "identity",  width = 0.4, position=position_dodge(width = 0.5)) + theme_classic() + labs(title = "Percentage of Homozygous Sites by Category", x = "Sample_ID", y = "Percent Homozygosity")

#Plot number of heterozygous sites of each sample by group 
ggplot(dataframe_maize_teosinte) + geom_point(mapping = aes(x=Sample_ID, y=Heterozygous, color=Group)) + labs(title = "Amount of Heterozygous Sites by Group", x = "Sample_ID", y = "Amount of Heterozygosity")
#Plot number of heterozygous site of each sample by category
ggplot(dataframe_maize_teosinte) + geom_point(mapping = aes(x=Sample_ID, y=Heterozygous, color=Category)) + labs(title = "Amount of Heterozygous Sites by Category", x = "Sample_ID", y = "Amount of Heterozygosity")

#Plot percentage of heterozgyous sites of each sample by group
ggplot(data = dataframe_maize_teosinte, aes(x = Sample_ID, y = Hetero_Percent, fill = Group )) + geom_bar(stat = "identity",  width = 0.4, position=position_dodge(width = 0.5)) + theme_classic() + labs(title = "Percentage of Heterozygous Sites by Group", x = "Sample_ID", y = "Percent Hetrozygosity")
#Plot percentage of heterozgyous sites of each sample by category
ggplot(data = dataframe_maize_teosinte, aes(x = Sample_ID, y = Hetero_Percent, fill = Category )) + geom_bar(stat = "identity",  width = 0.4, position=position_dodge(width = 0.5)) + theme_classic() + labs(title = "Percentage of Heterozygous Sites by Category", x = "Sample_ID", y = "Percent Heterozygosity")

#Plot proportion of homozygous to heterozygous sites of each sample by group
ggplot(data = dataframe_maize_teosinte, aes(x = Sample_ID, y = Homo_Hetero, fill = Group )) + geom_bar(stat = "identity",  width = 0.4, position=position_dodge(width = 0.5)) + theme_classic() + labs(title = "Proportion of Homozygous to Heterogygous Sites by Group", x = "Sample_ID", y = "Homozygous:Heterozygous Sites")
#Plot proportion of homozygous to heterozygous sites of each sample by category
ggplot(data = dataframe_maize_teosinte, aes(x = Sample_ID, y = Homo_Hetero, fill = Category )) + geom_bar(stat = "identity",  width = 0.4, position=position_dodge(width = 0.5)) + theme_classic() + labs(title = "Proportion of Homozygous to Heterogygous Sites by Category", x = "Sample_ID", y = "Homozygous:Heterozygous Sites")

#Plot amount of missing data of each sample by group 
ggplot(dataframe_maize_teosinte) + geom_point(mapping = aes(x=Sample_ID, y=Missing_Data, color=Group)) + labs(title = "Amount of Missing Data by Group", x = "Sample_ID", y = "Amount of Missing Data")
#Plot amount of missing data of each sample by category
ggplot(dataframe_maize_teosinte) + geom_point(mapping = aes(x=Sample_ID, y=Missing_Data, color=Category)) + labs(title = "Amount of Homozygous Sites by Category", x = "Sample_ID", y = "Amount of Missing Data")

#Plot percentage of missing data of each sample by group
ggplot(data = dataframe_maize_teosinte, aes(x = Sample_ID, y = Missing_Percent, fill = Group )) + geom_bar(stat = "identity",  width = 0.4, position=position_dodge(width = 0.5)) + theme_classic() + labs(title = "Percentage of Missing Data by Group", x = "Sample_ID", y = "Percent Missing Data")
#Plot percentage of missing data of each sample by category
ggplot(data = dataframe_maize_teosinte, aes(x = Sample_ID, y = Missing_Percent, fill = Category )) + geom_bar(stat = "identity",  width = 0.4, position=position_dodge(width = 0.5)) + theme_classic() + labs(title = "Percentage of Missing Data by Category", x = "Sample_ID", y = "Percent Missing Data")

#Visualizing homozygosity, heterozygosity and missing data as a whole by Group, Category and Sample:

#Create a data frame with five columns: "Sample_ID," "Group," and "Category" for both maize and teosinte, "SNP_ID" and whether the SNPs are heterozygous, homozygous, or containg missing data
fang_mt[fang_mt == "?/?"] <- "Missing"
fang_mt[fang_mt == "A/A" | fang_mt == "T/T" | fang_mt == "C/C" | fang_mt == "G/G"] <- "Homozygous"
fang_mt[fang_mt == "C/G" | fang_mt == "A/C" | fang_mt ==  "G/T" | fang_mt == "A/G" | fang_mt == "C/T" | fang_mt == "A/T"] <- "Heterozygous"
fang_mt$JG_OTU <- Group
colnames(fang_mt)[2] = "Group"
fang_mt_data <- fang_mt %>% pivot_longer(cols = -c("Sample_ID", "Category", "Group"), names_to = "SNP_ID", values_to = "SNP_Data")

#Plot fraction of homozygosity, heterozygosity and missing data by Group
ggplot(fang_mt_data) + geom_bar(mapping = aes(x=Group, fill = SNP_Data), position = "fill") + xlab("Group") + ylab("Fraction")
#Plot fraction of homozygosity, heterozygosity and missing data by Category
ggplot(fang_mt_data) + geom_bar(mapping= aes(x=Category, fill = SNP_Data), position = "fill") + xlab("Category") + ylab("Fraction")
#Plot fraction of homozygosity, heterozygosity and missing data by Sample
ggplot(fang_mt_data) + geom_bar(mapping = aes(x=Sample_ID, fill = SNP_Data), position = "fill") + xlab("Sample") + ylab("Fraction")

#Summary:

#When comparing the ratio of homozgyous to heterozygous sites in both maize and teosinte by sample, maize has more samples that have disproportionately higher homozygous SNP sites than teosinte
#When comparing both maize and teosinte, the amount of missing data tends to be similar between the two (less than 6.25%), with the amount of homozgyous SNP sites being higher than heterozygous sites, with only 12.5% and 25% of SNP_data being heterozygous, respectively

#Own Visualization

#Visualizing heterozygosity, homozygosity and missing data at each chromosome:

#Create a data frame with six columns: "Sample_ID," "Group," "Category," and "Chromosome" for both maize and teosinte, "SNP_ID" and whether the SNPs are heterozygous, homozygous, or containg missing data
Group <- fang_mt_data$Group
Category <- fang_mt_data$Category
SNP_ID <- fang_mt_data$SNP_ID
SNP_Data <- fang_mt_data$SNP_Data
Chromosome <- maize_teosinte_snp$Chromosome
Position <- maize_teosinte_snp$Position
fang_mt_own <- data.frame(Group, Category, Chromosome, Position, SNP_ID, SNP_Data)

#Plot fraction of homozygosity, heterozygosity and missing data at each chromosome
ggplot(fang_mt_own) + geom_bar(mapping = aes(x=Chromosome, fill = SNP_Data), position = "fill") + xlab("Chromosome") + ylab("Fraction")

#For chromosomes, the percentage of heterozygous sites, homozygous sites and missing data tends to be the same across, with missing data making up less than 6.25% and heterozygous sites make up about 12.5%
```